packages/tpetra/core/src/Tpetra_withLocalAccess.hpp

```diff
@@ -209,101 +253,327 @@ namespace Tpetra {
+    Details::read_write>
   readWrite (GlobalObjectType&);
 
   ////////////////////////////////////////////////////////////
   // LocalAccess struct
   ////////////////////////////////////////////////////////////
 
   namespace Details {
-    /// \brief Declaration of access intent for a global object.
+    //! Tag indicating an unspecified type in LocalAccessTraits.
+    struct unspecified_type {};
+
+    /// \brief Deduce types from parameter pack of LocalAccess.
     ///
-    /// Users aren't supposed to make instances of this class.  They
-    /// should use readOnly, writeOnly, or readWrite instead, then
-    /// call instance methods like on() and valid() on the resulting
-    /// LocalAccess instance.
+    /// Deduce the following types from Args:
+    /// <ul>
+    /// <li>execution_space (Kokkos execution space)</li>
+    /// <li>memory_space (Kokkos memory space)</li>
+    /// <li>access_mode (Access type)</li>
+    /// </ul>
+    template<class ... Args>
+    struct LocalAccessTraits {};
+
+    // An empty list of template arguments constrains nothing.
+    template<>
+    struct LocalAccessTraits<> {
+      using execution_space = unspecified_type;
+      using memory_space = unspecified_type;
+      using access_mode = unspecified_type;
+    };
+
+    // Skip over any instances of unspecified_type.  This makes it
+    // easy to define the return types of LocalAccess::on and
+    // LocalAccess::at (see below).
+    template<class ... Rest>
+    struct LocalAccessTraits<unspecified_type, Rest...> {
+    private:
+      using rest_type = LocalAccessTraits<Rest...>;
+    public:
+      using execution_space = typename rest_type::execution_space;
+      using memory_space = typename rest_type::memory_space;
+      using access_mode = typename rest_type::access_mode;
+    };
+
+    template<class First, class ... Rest>
+    struct LocalAccessTraits<First, Rest...> {
+    private:
+      using rest_type = LocalAccessTraits<Rest...>;
+    public:
+      using execution_space = typename std::conditional<
+        Kokkos::Impl::is_execution_space<First>::value,
+        First,
+        typename rest_type::execution_space>::type;
+      using memory_space = typename std::conditional<
+        Kokkos::Impl::is_memory_space<First>::value,
+        First,
+        typename rest_type::memory_space>::type;
+      using access_mode = typename std::conditional<
+        is_access_mode<First>::value,
+        First,
+        typename rest_type::access_mode>::type;
+    };
+
     template<class GlobalObjectType,
-             class MemorySpace,
-             const AccessMode am>
+             class Traits,
+             bool is_execution_space_specified = ! std::is_same<
+               typename Traits::execution_space,
+               unspecified_type>::value,
+             bool is_memory_space_specified = ! std::is_same<
+               typename Traits::memory_space,
+               unspecified_type>::value>
+    struct SpaceTypes {};
```

Comment:
We use this to select execution and memory spaces, based on whether the user specified either or both explicitly.